docs/lib/graphqlapi/fragments/ios/advanced-workflows.md

```diff
@@ -0,0 +1,225 @@
+
+```
+query getTodo($id: ID!) {
+  getTodo(id: $id) {
+    id
+    name
+  }
+}
+```
+The response data will look like this
+```json
+{
+  "data": {
+    "getTodo": {
+      "id": "111",
+      "name": "my first todo"
+    }
+  }
+}
+```
+First, create your own `GraphQLRequest`
+
+```swift
+extension GraphQLRequest {
+    static func getWithoutDescription(byId id: String) -> GraphQLRequest<Todo> {
+        let operationName = "getTodo"
+        let document = """
+        query getTodo($id: ID!) {
+          \(operationName)(id: $id) {
+            id
+            name
+          }
+        }
+        """
+        return GraphQLRequest<Todo>(document: document,
+                                    variables: ["id": id],
+                                    responseType: Todo.self,
+                                    decodePath: operationName)
+    }
+}
+```
+The decode path specifies which part of the response to deserialize to the `responseType`. You'll need to specify the operation name to deserialize the object at "data.getTodo" successfully into a Todo model.
+
+Then, query for the Todo by a todo id
+```swift
+Amplify.API.query(request: .getWithoutDescription(byId: "[UNIQUE_ID]")) { 
+  // handle result
+```  
+
+## Nested Data
+If you have a relational model, you can retrieve the nested object by creating a `GraphQLRequest` with a selection set containing the nested object's fields. For example, in this schema, the Post can contain multiple comments.
+```graphql
+enum PostStatus {
+  ACTIVE
+  INACTIVE
+}
+type Post @model {
+  id: ID!
+  title: String!
+  rating: Int!
+  status: PostStatus!
+  comments: [Comment] @connection(keyName: "byPost", fields: ["id"])
+}
+type Comment @model
+  @key(name: "byPost", fields: ["postID", "content"]) {
+  id: ID!
+  postID: ID!
+  post: Post! @connection(fields: ["postID"])
+  content: String!
+}
+```
+The `GraphQLRequest` will look like this
+
+```swift
+extension GraphQLRequest {
+    static func getPostWithComments(byId id: String) -> GraphQLRequest<JSONValue> {
+        let document = """
+        query getPost($id: ID!) {
+          getPost(id: $id) {
+            id
+            title
+            rating
+            status
+            comments {
+              items {
+                id
+                postID
+                content
+              }
+            }
+          }
+        }
+        """
+        return GraphQLRequest<JSONValue>(document: document,
+                                         variables: ["id": id],
+                                         responseType: JSONValue.self)
+    }
+}
+
+```
+Query with `Amplify.API.query(request: .getCommentWithPost(byId: "[COMMENT_ID]"))`. Notice here that `JSONValue` is used as the `responseType`. `JSONValue` is utility type that can be used to represent an arbitrary JSON response.
```

Comment:
Is there not a way for the customer to get a strongly typed response (i.e. a Post object and a list of Comment objects)?  For Android, the deserializer supports this, with an unlimited depth.

Reply:
The issue is first, the Model-to-GraphQLRequest builder does not generate the selection set of the comments so data is not returned. Even if we default to a depth of 2 and do add the selection set for comments, then the second issue is when deserializing it doesn't match the response data structure. The response data contains the post fields, followed by the "items" key, and inside "items" is the array of Comments. For a successfull deserialization, the type should be 
````
Post {
  var items: [Comment]
}
``` 
instead of 
Post {
  var comments: [Comment]
}
```
Does Android check for "items" and do some custom deserialization into the Post's list of comments model? Even if it does do a default selection set of depth 2, the list of comments should be paginated, how does the developer get the next x number of comments for that post?

Reply:
> The issue is first, the Model-to-GraphQLRequest builder does not generate the selection set of the comments so data is not returned. 

On Android, this is not an issue since we default to a depth of 2, for the API category.    To change the depth, all we need to do is expose a way to provide that constant.

> 
> Does Android check for "items" and do some custom deserialization into the Post's list of comments model?

Yes.  The expected type is `List<Comment>`, which is processed by the [`IterableDeserializer`](  https://github.com/aws-amplify/amplify-android/blob/27503d7a64c014e39eea4062d1962fb5b5128a27/aws-api/src/main/java/com/amplifyframework/api/aws/IterableDeserializer.java).  If the `JsonElement` is a `JsonObject`, rather than a `JsonArray`, (it is, and it contains `items` and `nextToken`), then the deserializer parses the value of `items` as a `List<Comment>`.

> Even if it does do a default selection set of depth 2, the list of comments should be paginated, how does the developer get the next x number of comments for that post?

Great question - Android doesn't support pagination on nested models, since we discard the pagination details during deserialization.   It is a limitation, but I don't think it's a huge deal because customers could still obtain a page of Comments by making a new request with Comments as the top level item, which does support pagination.

Reply:
So it is clear that Android and IOS contains different walk-depths by default, and it will take some major verions of the API to change this behavior. What i've done here is for the "nested data" section, instead of showcasing a walk-depth of two (this would be odd for Android since the default is already 2), it's showcasing the scenario where there's two nested objects and developer only wants to retrieve one of them. Post contains list of comments and and list of notes and only wants to get post with comments, omitting the notes